WireDomain/Sources/WireDomain/Synchronization/PushSupportedProtocolsUseCase.swift

```diff
@@ -0,0 +1,169 @@
+
+        /// All clients are proteus ready so we support it if the backend does.
+        if remoteProtocols.contains(.proteus) {
+            result.insert(.proteus)
+        }
+
+        /// All clients are mls ready so we support it if the backend does.
+        if remoteProtocols.contains(.mls), allClientsMLSReady {
+            result.insert(.mls)
+        }
+
+        /// Proteus is still supported if migration is pending or still ongoing.
+        if migrationState.isOne(of: .notStarted, .ongoing), allClientsMLSReady {
+            result.insert(.proteus)
+        }
+
+        /// MLS migration is complete.
+        if remoteProtocols.contains(.mls), migrationState == .finalised {
+            result.insert(.mls)
+        }
+
+        /// MLS is forced.
+        if remoteProtocols == [.mls], migrationState.isOne(of: .disabled, .finalised) {
+            result = [.mls]
+        }
+
+        /// Even if proteus isn't supported, migration is pending or still ongoing.
+        if remoteProtocols == [.mls], !allClientsMLSReady, migrationState.isOne(of: .notStarted, .ongoing) {
+            result = [.proteus]
+        }
+
+        logger.debug("calculated supported protocols: \(result)")
+
+        return result
+    }
+
+    private func remotelySupportedProtocols() async -> Set<WireAPI.MessageProtocol> {
+        let mlsFeature = try? await featureConfigRepository.fetchFeatureConfig(
+            with: .mls,
+            type: Feature.MLS.Config.self
+        )
+
+        let mls = (status: mlsFeature?.status ?? .disabled,
+                   config: mlsFeature?.config ?? Feature.MLS.Config())
+
+        guard mls.status == .enabled else {
+            /// If there is no MLS then there can only be proteus.
+            return [.proteus]
+        }
+
+        var result = Set<WireAPI.MessageProtocol>()
+
+        if mls.config.supportedProtocols.contains(.proteus) {
+            result.insert(.proteus)
+        }
+
+        if mls.config.supportedProtocols.contains(.mls) {
+            result.insert(.mls)
+        }
+
+        return result
+    }
+
+    private func currentMigrationState() async -> ProteusToMLSMigrationState {
+        let mlsMigrationFeature = try? await featureConfigRepository.fetchFeatureConfig(
+            with: .mlsMigration,
+            type: Feature.MLSMigration.Config.self
+        )
+
+        let mlsMigration = (status: mlsMigrationFeature?.status ?? .disabled,
+                            config: mlsMigrationFeature?.config ?? Feature.MLSMigration.Config())
+
+        guard mlsMigration.status == .enabled else {
+            return .disabled
+        }
+
+        let now = Date.now
+
+        guard
+            let startDate = mlsMigration.config.startTime,
+            startDate <= now
+        else {
+            return .notStarted
+        }
+
+        guard
+            let endDate = mlsMigration.config.finaliseRegardlessAfter,
+            endDate <= now
```

Comment:
Sanity check: Is this logic correct? I imagined that if now is after endDate we should already be finalized.

Reply:
it looks like `finaliseRegardlessAfter` is the deadline, if we have not reached this date yet, the migration is still ongoing, so theoretically, the above logic is incorrect (inversed actually) however this is the legacy code so I assume there is some reason we are doing so

Reply:
will double check on Android, out of curiosity

Reply:
on Android, logic is the same

```
        fun MLSMigrationModel.hasMigrationEnded(): Boolean {
            return status == Status.ENABLED && endTime?.let { Clock.System.now() > it } ?: false
        }
```